pkg/compiler/check.go

```diff
@@ -954,7 +954,7 @@ func (comp *compiler) instantiate(templ ast.Node, params []*ast.Ident, args []*a
 	}
 	argUsed := make(map[string]bool)
 	err0 := comp.errors
-	templ.Walk(ast.PostRecursive(func(n ast.Node) {
+	recurCheck := func(n ast.Node) {
```

Comment:
I would assume that Walk would visit the top node as well. What if you make pkg/ast/walk.go work this way? Does it break something else? Still it feels that the right way to fix it.
I also wonder if there is more elegant way to fix this than adding cb(n) to every Walk implementation?...  Perhaps we need to unexport Node.Walk and provide wrapper functions that visit the current node and do Pre/Post recursive walk of child nodes. Do we use non-recursive walk anywhere?

Reply:
> I would assume that Walk would visit the top node as well. What if you make pkg/ast/walk.go work this way? Does it break something else?

Yes, it would break `PostRecursive(cb)`.

> I also wonder if there is more elegant way to fix this than adding cb(n) to every Walk implementation?... Perhaps we need to unexport Node.Walk and provide wrapper functions that visit the current node and do Pre/Post recursive walk of child nodes.

Definitely. That way we can make it work for `PostRecursive(cb)`. Any idea of names for the wrapper functions? What about `[Pre/Post]Traversal(node)`?

> Do we use non-recursive walk anywhere?

Not as far as I could see. `compiler.instantiate()` seems the only place calling `Node.Walk()`.

Reply:
I don't understand why PostRecursive breaks, but we can name them Walk as well, just a global function.

Reply:
> I don't understand why PostRecursive breaks

It wouldn't if we were to define a `Node.Walk()` *and* a `Node.PostWalk()`. But `Recursive()` and `PostRecursive()` can't be calling the same `Node.Walk()` method if that method is itself calling `cb(n)`.

Anyway, it's cleaner with wrappers. Ok for `Walk()` and `PostWalk()` then (to be consistent with `Recursive()` and `PostRecursive()`).

Reply:
Any reason we can't simply call the function returned by `PostRecursive` on the root node? See [my `fix-pre-post-walks` branch](https://github.com/pchaigno/syzkaller/commit/4a425fd32c16b84dcfb7763ac99ed8a86a9a12b8#diff-47a4f07e1880cd23bc05dca73d329c1d) which implements that change (+unexport `Node.Walk()`).

That way the root node is visited, we don't need global wrapper functions, and we can unexport `Node.Walk()`.

Reply:
Let's do this.
I just want to ensure that I won't do the same bug again tomorrow :)
If walk is not exported, nobody will call it directly.

Reply:
Pull request opened at #1475.